ably/realtime_channel.go

```diff
@@ -167,9 +167,37 @@ func (c *RealtimeChannel) mayAttach(result, checkActive bool) (Result, error) {
 			return nopResult, nil
 		}
 	}
-	if !c.client.Connection.lockIsActive() {
-		return nil, c.state.set(StateChanFailed, errAttach)
+
+	switch c.client.Connection.State() {
+	// RTL4b
+	case StateConnClosed,
+		StateConnClosing,
+		StateConnFailed:
+		return nil, newError(80000, errAttach)
+
+	// RTL4i
+	case StateConnInitialized,
+		StateConnConnecting,
+		StateConnDisconnected:
+
+		return goWaiter(func() error {
+			connected := make(chan State, 1)
+			c.client.Connection.On(connected, StateConnConnected)
+			<-connected
```

Comment:
This will block forever when `StateConnConnected` isn't fired.

Reply:
Good catch! Fixed at 71e958c.